internal/pkg/service/buffer/sink/tablesink/storage/repository/file_repo_test.go

```diff
@@ -351,94 +318,1056 @@ storage/file/level/target/123/456/my-source/my-sink-1/2000-01-01T19:00:00.000Z
+	file1Volume2Key := storage.FileVolumeKey{FileKey: file1.FileKey, VolumeID: volumeID2}
+	{
+		clk.Add(time.Hour)
+		require.NoError(t, sliceFacade.Rotate(clk.Now(), file1Volume2Key).Do(ctx).Err())
+	}
 
-	// Create 3 files
-	clk.Add(time.Hour)
-	fileKey1 := storage.FileKey{SinkKey: sinkKey, FileID: storage.FileID{OpenedAt: utctime.From(clk.Now())}}
-	require.NoError(t, r.Create(fileKey1, credentials).Do(ctx).Err())
-	clk.Add(time.Hour)
-	fileKey2 := storage.FileKey{SinkKey: sinkKey, FileID: storage.FileID{OpenedAt: utctime.From(clk.Now())}}
-	require.NoError(t, r.Create(fileKey2, credentials).Do(ctx).Err())
-	clk.Add(time.Hour)
-	fileKey3 := storage.FileKey{SinkKey: sinkKey, FileID: storage.FileID{OpenedAt: utctime.From(clk.Now())}}
-	require.NoError(t, r.Create(fileKey3, credentials).Do(ctx).Err())
+	// Rotate file (2)
+	// -----------------------------------------------------------------------------------------------------------------
+	{
+		clk.Add(time.Hour)
+		file2, err := fileFacade.Rotate(rb, clk.Now(), sinkKey).Do(ctx).ResultOrErr()
+		require.NoError(t, err)
+		assert.Equal(t, clk.Now(), file2.OpenedAt().Time())
+	}
+
+	// Rotate file (3)
+	// -----------------------------------------------------------------------------------------------------------------
+	var file3 storage.File
+	{
+		var err error
+		clk.Add(time.Hour)
+		file3, err = fileFacade.Rotate(rb, clk.Now(), sinkKey).Do(ctx).ResultOrErr()
+		require.NoError(t, err)
+		assert.Equal(t, clk.Now(), file3.OpenedAt().Time())
+	}
+
+	// Create slice3 in the file, in the volume1
+	// -----------------------------------------------------------------------------------------------------------------
+	file3Volume1Key := storage.FileVolumeKey{FileKey: file3.FileKey, VolumeID: volumeID1}
+	{
+		clk.Add(time.Hour)
+		require.NoError(t, sliceFacade.Rotate(clk.Now(), file3Volume1Key).Do(ctx).Err())
+	}
 
-	// Old file is always switched from the FileWriting state, to the FileClosing state
+	// Create slice4 in the file, in the volume2
+	// -----------------------------------------------------------------------------------------------------------------
+	file3Volume2Key := storage.FileVolumeKey{FileKey: file3.FileKey, VolumeID: volumeID2}
+	{
+		clk.Add(time.Hour)
+		require.NoError(t, sliceFacade.Rotate(clk.Now(), file3Volume2Key).Do(ctx).Err())
+	}
+
+	// Check Storage API calls
+	// -----------------------------------------------------------------------------------------------------------------
+	// File prepare endpoint should be called N times
+	assert.Equal(t, 3, transport.GetCallCountInfo()["POST /v2/storage/branch/456/files/prepare"])
+	assert.Equal(t, 0, transport.GetCallCountInfo()[`DELETE =~/v2/storage/branch/456/files/\d+$`])
+
+	// Test rollback, delete file endpoint should be called N times
+	rb.Invoke(ctx)
+	assert.Equal(t, 3, transport.GetCallCountInfo()[`DELETE =~/v2/storage/branch/456/files/\d+$`])
```

Comment:
We can easily test rollback on fail, together with the sucessful operation, in one test.

Rollback is in production triggered in the top - API layer - where the `rb` instance is created.
https://github.com/keboola/keboola-as-code/blob/c10891eb0402162d1abab70b952cb1ca627bba49/internal/pkg/service/buffer/api/service/export.go#L63-L65